### ISSUE TITLE
ColorRect flickers when shown by mouse_entered signal

### ISSUE THREAD

**Issue by drmar** [godotengine/godot#35922]
**Godot version:**
3.2 Stable

**OS/device including version:**
macOS Catalina 10.15.3 / MacBookPro11,1

**Issue description:**
When calling show() inside a function on a ColorRect node, from mouse_entered signals, the ColorRect node flickers.

**Steps to reproduce:**
Use the Godot 3.2 project attached.

**Minimal reproduction project:**
[Godot 3.2.zip](https://github.com/godotengine/godot/files/4158725/Godot.3.2.zip)

Video:
[mouse_entered_flicker.mov.zip](https://github.com/godotengine/godot/files/4158730/mouse_entered_flicker.mov.zip)

**Comment by drmar** [comment]
I believe that the signals mouse_entered and mouse_exited are called very rapidly successively, even though the mouse never left the CollisionShape2D.

**Comment by Torguen** [comment]
Please provide the necessary information, for example:
You must open the "Game.tscn" scene and press "Space", then bring the cursor to the ax icon.

What is the script with the code related to the error? There are many scripts in the "Game.tscn" scene.

**Comment by drmar** [comment]
There are only 2 scripts that have relevance:

- Tool_Weapon_Pickaxe.gd
- UI.gd

All you have to do is run the project as is, press Space to open the Tool menu and hover the mouse over the Pickaxe sprite - which has an Area2D with CollisionShape2D attached.

The results can be seen in the video I attached.

**Comment by drmar** [comment]
Here is the same script being used in a minimal similar project made in Godot 3.1.2 that doesn't have this problem:

[Dig and Build.zip](https://github.com/godotengine/godot/files/4160273/Dig.and.Build.zip)

**Comment by kleonc** [comment]
To fix your problem:
open "Pickaxe.tscn", select "bg" ColorRect, in the inspector in section "Mouse" set "Filter" to "Ignore".

What seems to be happening now:
- Area2D detects mouse and emits mouse_entered signal (and you are showing ColorRect in the handler)
- after showing ColorRect it is by default handling future mouse events over it,
- so Area2D is no longer detecting mouse over it,
- so Area2D emits mouse_exited signal (and you are hiding ColorRect in the handler),
- after hiding ColorRect it is no longer handling future mouse events so Area2D will detect mouse_entered after next mouse move,

and these steps will keep looping (as long as you are moving mouse over them).

Also you should probably consider using TextureRect (or some other Control) instead of Sprite + Area2D combination. TextureRect is a Control and controls are designed to be used in GUI.

**Comment by drmar** [comment]
@kleonc, did you check the same project in Godot 3.1.2?

It uses the same exact script and it doesn't cause the problem as in Godot 3.2.

Did something fundamental change between Godot 3.1.2 and Godot 3.2 to the way that mouse_eneterd and mouse_exited works?

**Comment by kleonc** [comment]
@drmar, yes I checked it and i can confirm your issue, same thing happening to me.

I've created a little simpler reproduction project: 
[FlickeringTest.zip](https://github.com/godotengine/godot/files/4164961/FlickeringTest.zip)

Scene.gd:
```
func _on_Area2D_mouse_entered():
	print("_on_Area2D_mouse_entered")
	$ColorRect.show()

func _on_Area2D_mouse_exited():
	print("_on_Area2D_mouse_exited")
	$ColorRect.hide()

func _on_ColorRect_mouse_entered():
	print("_on_ColorRect_mouse_entered")

func _on_ColorRect_mouse_exited():
	print("_on_ColorRect_mouse_exited")

func _on_ColorRect_gui_input(event):
	print("_on_ColorRect_gui_input, event = ", event)

func _on_Area2D_input_event(viewport, event, shape_idx):
	print("_on_Area2D_input_event, event = ", event)
```

1. Running in Godot_v3.1.2-stable_win64.exe:
- ColorRect mouse filter Stop
![FlickeringTest_godot_3 1 2_small](https://user-images.githubusercontent.com/9283098/73935872-d82eab00-48e1-11ea-845e-9bc78aee2101.gif)
- ColorRect mouse filter Ignore
![FlickeringTest_godot_3 1 2_filterIgnore_small](https://user-images.githubusercontent.com/9283098/73936513-517acd80-48e3-11ea-9a17-d189b42a8280.gif)



2. Running in Godot_v3.2-stable_win64.exe:
- ColorRect mouse filter Stop
![FlickeringTest_godot_3 2_small](https://user-images.githubusercontent.com/9283098/73935890-e250a980-48e1-11ea-80d1-cd36a9647357.gif)
- ColorRect mouse filter Ignore
![FlickeringTest_godot_3 2_filterIgnore_small](https://user-images.githubusercontent.com/9283098/73936517-550e5480-48e3-11ea-9b30-7cd77b906d36.gif)


I don't know what causes this issue nor if it is a bug or an expected behaviour.

**Comment by adomas-skliutas** [comment]
Having a maybe related issue with Godot Engine v3.3.1 GLES3. I have a Control node with ColorRect and Label as children and signals for mouse entered/exited that sets ColorRect color. after color change the Label gets distorted.
extends Control.

```
func _on_DroppedItem_mouse_entered():
	$ColorRect.color = Color(0, 0, 0, 120)

func _on_DroppedItem_mouse_exited():
	$ColorRect.color = Color(0, 0, 0, 200)
```
Before color change
<img width="303" alt="image" src="https://user-images.githubusercontent.com/23510046/119190912-a0bc2180-ba86-11eb-8b8d-6a9f31baaccb.png">

After color change
<img width="194" alt="image" src="https://user-images.githubusercontent.com/23510046/119190941-a87bc600-ba86-11eb-8043-5f41a59a2379.png">

**Comment by Calinou** [comment]
@adomas-skliutas Can you reproduce this after disabling batching in the Project Settings?

**Comment by adomas-skliutas** [comment]
@Calinou If you mean Rendering/Batching/Use Batching -> false, then yes, can still reproduce

**Comment by KoBeWi** [comment]
No flickering in 4.4
